cpp/src/umap/umap.cu

```diff
@@ -15,117 +15,115 @@
  */
 
 #include <cuml/manifold/umapparams.h>
+#include <cuml/manifold/common.hpp>
 #include <cuml/manifold/umap.hpp>
 #include "runner.cuh"
 
 #include <iostream>
 
 namespace ML {
+namespace UMAP {
 
-static const int TPB_X = 256;
+static const int TPB_X = 32;
 
+// Dense transform
 void transform(const raft::handle_t &handle, float *X, int n, int d,
-               int64_t *knn_indices, float *knn_dists, float *orig_X,
-               int orig_n, float *embedding, int embedding_n,
+               knn_indices_dense_t *knn_indices, float *knn_dists,
+               float *orig_X, int orig_n, float *embedding, int embedding_n,
                UMAPParams *params, float *transformed) {
-  UMAPAlgo::_transform<float, TPB_X>(handle, X, n, d, knn_indices, knn_dists,
-                                     orig_X, orig_n, embedding, embedding_n,
-                                     params, transformed);
+  if (knn_indices != nullptr && knn_dists != nullptr) {
+    manifold_precomputed_knn_inputs_t<knn_indices_dense_t, float> inputs(
+      knn_indices, knn_dists, X, nullptr, n, d, params->n_neighbors);
+    UMAPAlgo::_transform<
+      knn_indices_dense_t, float,
+      manifold_precomputed_knn_inputs_t<knn_indices_dense_t, float>, TPB_X>(
+      handle, inputs, inputs, embedding, embedding_n, params, transformed);
+  } else {
+    manifold_dense_inputs_t<float> inputs(X, nullptr, n, d);
+    manifold_dense_inputs_t<float> orig_inputs(orig_X, nullptr, orig_n, d);
+    UMAPAlgo::_transform<knn_indices_dense_t, float,
+                         manifold_dense_inputs_t<float>, TPB_X>(
+      handle, inputs, orig_inputs, embedding, embedding_n, params, transformed);
+  }
 }
+
+// Sparse transform
+void transform_sparse(const raft::handle_t &handle, int *indptr, int *indices,
+                      float *data, size_t nnz, int n, int d, int *orig_x_indptr,
+                      int *orig_x_indices, float *orig_x_data, size_t orig_nnz,
+                      int orig_n, float *embedding, int embedding_n,
+                      UMAPParams *params, float *transformed) {
+  manifold_sparse_inputs_t<knn_indices_sparse_t, float> inputs(
+    indptr, indices, data, nullptr, nnz, n, d);
+  manifold_sparse_inputs_t<knn_indices_sparse_t, float> orig_x_inputs(
+    orig_x_indptr, orig_x_indices, orig_x_data, nullptr, orig_nnz, orig_n, d);
+
+  UMAPAlgo::_transform<knn_indices_sparse_t, float,
+                       manifold_sparse_inputs_t<int, float>, TPB_X>(
+    handle, inputs, orig_x_inputs, embedding, embedding_n, params, transformed);
+}
```

Comment:
I am under the impression that if the training data is dense, the data to be transformed has to be dense as well and conversely if it is sparse. There's probably a good reason for that... Wouldn't it be possible to run the KNN and the rest of the algorithm with inputs of different sorts (dense/sparse)?

Reply:
Unfortunately, `transform()` requires a KNN query between the transform input and the original training input, which creates a coupling between the two formats. I think in the future we might be able to use a GEMM that supports sparse & dense inputs, but the current priority is a standard sparse KNN and enabling UMAP to support sparse inputs. I'll create an issue for this to gather user feedback as well.